UCP/API: Add CPU affinity support to UCP worker

guy101 opened the pull request:
UCP/API: Added  worker_params value to ucp_worker_create()  …
               Added cpu_set to worker_params

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2644/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/45/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2645/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/46/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2001/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2000/ for details (Mellanox internal link).

shamisp commented:
@yosefe - maybe we should go with hwloc ?

yosefe commented:
@shamisp how would you suggest to modify this PR to use hwloc?

shamisp commented:
@yosefe - we should to discuss all the options. I'm not sure if I want to expose HWLOC dependency on the API level. My understanding that CPUSET is not very consistent across various versions of linux kernel and for sure is not portable across OS's

guy101 commented:
bot:retest

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2646/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/47/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2002/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/48/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2003/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2648/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2649/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/50/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2005/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/51/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2650/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2652/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/53/ for details.

guy101 commented:
@yosefe  build finally passed successfully. please code-review

yosefe commented on src/tools/info/proto_info.c:
```diff
@@ -43,7 +43,9 @@ void print_ucp_info(int print_opts, ucs_config_print_flags_t print_flags,
 
-    status = ucp_worker_create(context, UCS_THREAD_MODE_MULTI, &worker);
+    ucp_worker_params_t ucp_worker_params;
+    bzero(&ucp_worker_params,sizeof(ucp_worker_params));
```
- bzero is deprecated
- code style

guy101 replied:
done

yosefe commented on src/tools/info/proto_info.c:
```diff
@@ -43,7 +43,9 @@ void print_ucp_info(int print_opts, ucs_config_print_flags_t print_flags,
-    status = ucp_worker_create(context, UCS_THREAD_MODE_MULTI, &worker);
+    ucp_worker_params_t ucp_worker_params;
+    bzero(&ucp_worker_params,sizeof(ucp_worker_params));
+    status = ucp_worker_create(context, &ucp_worker_params, UCS_THREAD_MODE_MULTI, &worker);
```
need to add thread_mode to ucp_worker_params_t

guy101 replied:
done

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -13,6 +13,10 @@
 #include <ucs/sys/math.h>
 #include <stdio.h>
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
```
this is dangerous thing in public API, we should define a structure which is platform-independent
@shamisp

guy101 replied:
I will update the code as soon as you have a resolution

guy101 replied:
TODO

shamisp replied:
this is a really bad idea. We have to figure out how to handle this. I don't want to API level dependance on HWLOC either. This pretty much leaves us only with the option of defining own interface for cpu query ...

yosefe replied:
@guy101 pls remove this

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -168,6 +172,10 @@ enum ucp_dt_type {
 };
 
+struct ucp_worker_params {
+    cpu_set_t                     cpu_set; /* cpu_set affinity to use when creating a cq */
```
* pls use doxygen style comments and document the structure as well
* add field_mask for compatibility, see ucp_context_params_t

guy101 replied:
I will change it once a new issue is opened and a new PR is completed - about parameters

yosefe commented on src/ucp/api/ucp_def.h:
```diff
@@ -43,6 +43,15 @@ typedef struct ucp_tag_recv_info         ucp_tag_recv_info_t;
+ * UCP worker params is a set of parameters passed to ucp_worker_create()
+ * upon creation of ucp worker
+ */
+
```
no need for space line here

guy101 replied:
done

yosefe commented on src/ucp/api/ucp_def.h:
```diff
@@ -43,6 +43,15 @@ typedef struct ucp_tag_recv_info         ucp_tag_recv_info_t;
 typedef struct ucp_context               *ucp_context_h;
 
+/**
+ * @ingroup UCP_WORKER_PARAMS
```
where was the group UCP_WORKER_PARAMS defined?

guy101 replied:
it was not defined. we need to talk so you can explain me when and how to define those

guy101 replied:
TODO

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -171,7 +171,8 @@ static void ucp_worker_wakeup_context_cleanup(ucp_worker_wakeup_t *wakeup)
 static ucs_status_t ucp_worker_add_iface(ucp_worker_h worker,
-                                         ucp_rsc_index_t tl_id)
+                                         ucp_rsc_index_t tl_id,
+                                         cpu_set_t cq_cpu_set)
```
this is large value, please pass as const*

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -193,6 +194,7 @@ static ucs_status_t ucp_worker_add_iface(ucp_worker_h worker,
     iface_params.tl_name     = resource->tl_rsc.tl_name;
     iface_params.dev_name    = resource->tl_rsc.dev_name;
     iface_params.rx_headroom = sizeof(ucp_recv_desc_t);
+    iface_params.cq_cpu_set  = cq_cpu_set;
```
'cq' terminology is hw specific

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -380,8 +381,21 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
     return UCS_OK;
 }
 
+static int find_lcs_cpu(cpu_set_t const *cpu_cores)
```
name convention

guy101 replied:
done

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -380,8 +381,21 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
+static int find_lcs_cpu(cpu_set_t const *cpu_cores)
+{
+    int i;
+    for(i = 0 ; i < CPU_COUNT(cpu_cores); ++i) {
```
code style

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -380,8 +381,21 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
+{
+    int i;
+    for(i = 0 ; i < CPU_COUNT(cpu_cores); ++i) {
+        if(CPU_ISSET(i, cpu_cores)) {
```
code style

guy101 replied:
done

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -516,13 +530,14 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
     }
 
-    status = uct_ib_iface_create_cq(self, tx_cq_len, 0, &self->send_cq);
+    int cq_comp_vector = find_lcs_cpu(&params->cq_cpu_set);
```
no c99 please

guy101 replied:
:-(

done

yosefe commented on test/gtest/uct/uct_test.cc:
```diff
@@ -209,10 +209,12 @@ uct_test::entity::entity(const resource& resource, uct_iface_config_t *iface_con
 
     uct_iface_params_t iface_params;
-
+    cpu_set_t          cq_cpu_set;
```
no need for extra var, can memset(&iface_params.cpu_set, ...)

guy101 replied:
done

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2006/ for details (Mellanox internal link).

guy101 commented:
@yosefe   I addresed most of your comments. please read my replies.

yosefe commented on src/tools/info/proto_info.c:
```diff
@@ -43,7 +43,10 @@ void print_ucp_info(int print_opts, ucs_config_print_flags_t print_flags,
     }
 
-    status = ucp_worker_create(context, UCS_THREAD_MODE_MULTI, &worker);
+    ucp_worker_params_t ucp_worker_params;
```
c99

guy101 replied:
done

yosefe commented on src/tools/perf/libperf.c:
```diff
@@ -1012,7 +1012,10 @@ static ucs_status_t ucp_perf_setup(ucx_perf_context_t *perf, ucx_perf_params_t *
     }
 
-    status = ucp_worker_create(perf->ucp.context, params->thread_mode,
+    ucp_worker_params_t ucp_worker_params;
```
c99

yosefe commented on src/tools/perf/libperf.c:
```diff
@@ -1012,7 +1012,10 @@ static ucs_status_t ucp_perf_setup(ucx_perf_context_t *perf, ucx_perf_params_t *
 
-    status = ucp_worker_create(perf->ucp.context, params->thread_mode,
+    ucp_worker_params_t ucp_worker_params;
+    bzero(&ucp_worker_params,sizeof(ucp_worker_params));
```
bzero deprecated

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -168,6 +172,11 @@ enum ucp_dt_type {
 };
 
+struct ucp_worker_params {
+    ucs_thread_mode_t             thread_mode;
```
need doxygen

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -168,6 +172,11 @@ enum ucp_dt_type {
 
+struct ucp_worker_params {
+    ucs_thread_mode_t             thread_mode;
+    cpu_set_t                     cpu_set; /* cpu_set affinity to use when creating a cq */
```
need doxygen style comment

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -435,7 +437,7 @@ unsigned ucp_worker_get_ep_config(ucp_worker_h worker,
 }
 
-ucs_status_t ucp_worker_create(ucp_context_h context, ucs_thread_mode_t thread_mode,
+ucs_status_t ucp_worker_create(ucp_context_h context, ucp_worker_params_t *worker_params,
```
should be new function, and worker_params should be const

guy101 replied:
will be done as part of a seperate issue/PR. as discussed

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -380,8 +381,21 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
+    return 0;
+}
+
+
```
need 1 space line here, not 2

guy101 replied:
done

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -380,8 +381,21 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
+static int uct_ib_find_lcs_cpu(cpu_set_t const *cpu_cores)
+{
+    int i;
+    for(i = 0; i < CPU_COUNT(cpu_cores); ++i) {
```
code style

guy101 replied:
please elaborate

yosefe commented on test/examples/ucp_hello_world.c:
```diff
@@ -423,7 +423,10 @@ int main(int argc, char **argv)
     }
 
-    status = ucp_worker_create(ucp_context, UCS_THREAD_MODE_SINGLE, &ucp_worker);
+    ucp_worker_params_t worker_params;
```
c99

guy101 replied:
done

yosefe commented on test/examples/ucp_hello_world.c:
```diff
@@ -423,7 +423,10 @@ int main(int argc, char **argv)
 
-    status = ucp_worker_create(ucp_context, UCS_THREAD_MODE_SINGLE, &ucp_worker);
+    ucp_worker_params_t worker_params;
+    bzero(&worker_params, sizeof(ucp_worker_params_t));
```
no bzero pls

guy101 replied:
done

yosefe commented on test/gtest/ucp/ucp_test.cc:
```diff
@@ -257,9 +257,11 @@ ucp_test_base::entity::entity(const ucp_test_param& test_param, ucp_config_t* uc
                            &test_param.ctx_params, ucp_config);
-
+    ucp_worker_params_t worker_params;
+    bzero(&worker_params, sizeof(ucp_worker_params_t));
```
no bzero

guy101 replied:
done

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2008/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2655/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/57/ for details.

guy101 commented:
@yosefe re-review please

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2656/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2657/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/58/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2012/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2011/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2013/ for details (Mellanox internal link).

guy101 commented:
bot:retest

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/60/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/61/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2659/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2660/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2661/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2662/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/62/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2017/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2016/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2015/ for details (Mellanox internal link).

guy101 commented:
bot:retest

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/63/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/64/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2663/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2664/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2665/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2666/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/65/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2018/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2019/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2667/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/66/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2020/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2021/ for details (Mellanox internal link).

guy101 commented:
@yosefe I addressed all the comments.
- compilation and tests passed.
- The only failed tests relate to missing head info regarding the cpu_set_t structures and macros.
please suggest how to proceed.

yosefe commented:
@guy101 long long could be too small 
pls create datatype similar to cpu_set_t in size

guy101 commented:
Apparently it is bit-array.
I'll create a similar bit array with different name

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2677/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2037/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2678/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2679/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2038/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2039/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2680/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2681/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2682/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2040/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2041/ for details (Mellanox internal link).

yosefe commented on src/tools/perf/libperf.c:
```diff
@@ -912,6 +912,7 @@ static ucs_status_t uct_perf_setup(ucx_perf_context_t *perf, ucx_perf_params_t *
         .dev_name    = params->uct.dev_name,
         .rx_headroom = 0
     };
+    memset(&iface_params.cq_cpu_set, 0, sizeof(iface_params.cq_cpu_set));
```
pls use new UCS_CPU_ZERO macro

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -170,6 +175,42 @@ enum ucp_dt_type {
+      *  thread mode for ucp worker
+      */
+    ucs_thread_mode_t             thread_mode;
+     /**
```
pls add space between lines

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -639,6 +680,32 @@ void ucp_context_print_info(ucp_context_h context, FILE *stream);
+ *
+ * @return Error code as defined by @ref ucs_status_t
+ */
+ucs_status_t ucp_worker_create_with_params(ucp_context_h context, const ucp_worker_params_t *worker_params,
```
line too long

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,73 @@
+#include <stddef.h>
+
+/* Type for array elements in 'cpu_set_t'.  */
+typedef unsigned long int ucs_cpu_mask;
```
pls fix code style in this file, remove unused macros, and re-write anything that was copied from GPL headers

guy101 replied:
TODO

guy101 replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -17,10 +17,18 @@
 
 #include <sys/socket.h>
 #include <stdio.h>
+#ifndef _GNU_SOURCE
```
pls remove this

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -380,8 +382,20 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
     return UCS_OK;
 }
 
+static int uct_ib_find_lcs_cpu(ucs_cpu_set_t const *cpu_cores)
```
we should move this to ucs/type/cpu_set.h

guy101 replied:
agree

yosefe replied:
what problem? this function is static

guy101 replied:
/home/guy/work/ucx.cq/ucx/src/ucs/type/cpu_set.h:71:12: error: 'uct_ib_find_lcs_cpu' defined but not used [-Werror=unused-function]

yosefe replied:
static inline

guy101 replied:
@yosefe static inline does not solve the problem 'defined but not used'

guy101 replied:
only solution is moving leaving it 'as is' or moving it to cpu_set.c and adding it to Makefile

yosefe commented on test/gtest/uct/uct_test.cc:
```diff
@@ -8,6 +8,7 @@
 
 extern "C" {
 #include <ucs/stats/stats.h>
+#include <ucs/type/cpu_set.h>
```
no need to include this file

guy101 replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -318,6 +326,7 @@ struct uct_iface_attr {
  * @ref uct_iface_open. User has to initialize all fields of this structure.
  */
 struct uct_iface_params {
+    ucs_cpu_set_t            cq_cpu_set;  /** <cpu set to use while creating cq */
```
cq is ib specific, pls remove

yosefe commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -11,10 +11,12 @@
 #include <string.h>
 #include <stdlib.h>
 #include <poll.h>
+#include <sched.h>
```
is this needed?

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2683/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2042/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2043/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/80/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/81/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/82/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/83/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/84/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/85/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/86/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2686/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/89/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2046/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2688/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/91/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2048/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2691/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2692/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/94/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/95/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/96/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2693/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2694/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2695/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/97/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2055/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2054/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2697/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/99/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2056/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2057/ for details (Mellanox internal link).

guy101 commented:
bot:retest

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2698/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/100/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2059/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2060/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2699/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/101/ for details.

shamisp commented:
Why do we have thread mode in this PR ?

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2061/ for details (Mellanox internal link).

guy101 commented:
please check ucp_worker_params, it was there before. I merely put ucs_cpu_set in the beginning of the struct.

shamisp commented:
@guy101 in the github diff it appears as a new change. My educated guess - you branch is based out of some un-merged PR. If this is the case then we have to wait until we merge the other branch. Please double check with @yosefe . thanks

yosefe commented:
@shamisp we moved thread_mode inside ucp_worker_params and added a corresponding bit to field_mask

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -438,6 +440,18 @@ unsigned ucp_worker_get_ep_config(ucp_worker_h worker,
+	/* set default parameters for ucp worker */
+    worker_params.field_mask  = UCP_PARAM_FIELD_THREAD_MODE | UCP_PARAM_FIELD_CPU_SET;
+    worker_params.thread_mode = thread_mode;
+    memset(&worker_params.cpu_set, 0, sizeof(worker_params.cpu_set));
```
better to not pass UCP_PARAM_FIELD_THREAD_MODE in the first place

guy101 replied:
what do you want me to do about it?

yosefe replied:
avoid the memset(), and set field_mask to UCP_PARAM_FIELD_THREAD_MODE only.

guy101 replied:
done

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -496,7 +510,7 @@ ucs_status_t ucp_worker_create(ucp_context_h context, ucs_thread_mode_t thread_m
 
     /* Create the underlying UCT worker */
-    status = uct_worker_create(&worker->async, thread_mode, &worker->uct);
+    status = uct_worker_create(&worker->async, worker_params->thread_mode, &worker->uct);
```
pls consider field_mask first

yosefe replied:
you have to check params->field_mask before accessing every field - that was the purpose of adding field_mask. if the bit is not set, need to use a default value.

guy101 replied:
what should the default value be?

yosefe replied:
thread mode: UCS_THREAD_MODE_SINGLE
cpuset : all 0-s

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,83 @@
+#include <stddef.h>
+
+/* Type for array elements in 'ucs_cpu_set_t'.  */
+typedef unsigned long int ucs_cpu_mask;
```
bad name for type

guy101 replied:
advise other name

yosefe replied:
per https://github.com/openucx/ucx/blob/master/doc/CodeStyle#L21:
ucs_cpu_mask_t

guy101 replied:
done

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,83 @@
+
+
+/* Size definition for CPU sets.  */
+# define UCS_CPU_SETSIZE  1024
```
why space after #

guy101 replied:
removed

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,83 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+* Copyright (C) The University of Tennessee and The University 
```
AFAIK UoT did not write this file

guy101 replied:
removed

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,83 @@
+# define UCS_CPU_SETSIZE  1024
+# define UCS_NCPUBITS     (8 * sizeof (ucs_cpu_mask))
+
+# define UCS_CPUELT(cpu)  ((cpu) / UCS_NCPUBITS)
```
@guy101 pls fix code style in this file and remove unused functions

guy101 replied:
code cleaned up. this is the final result.

shamisp commented:
I would say this has to go as a separate PR, unless you have strong feeling
about it :)

On Wed, Oct 26, 2016 at 10:49 AM, Yossi <notifications@github.com> wrote:

> @shamisp <https://github.com/shamisp> we moved thread_mode inside
> ucp_worker_params and added a corresponding bit to field_mask
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openucx/ucx/pull/1070#issuecomment-256391316>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACIe2JluwQWkMJT9e2_MQXMnF_plwP7lks5q33aDgaJpZM4KcDGi>
> .
>

guy101 commented:
@yosefe  please collect all your comments at once so I can submit one (or two) major corrections. 
I'm tired of correcting minor syntax/style issues again and again and again.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2703/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2065/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/111/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2704/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/112/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2066/ for details (Mellanox internal link).

shssf commented on src/ucp/api/ucp.h:
```diff
@@ -170,6 +171,44 @@ enum ucp_dt_type {
+      *  thread mode for ucp worker
+      */
+    ucs_thread_mode_t             thread_mode;
+};
```
Do we really want to define `ucp_worker_params_t` in other `ucp/api/ucp_def.h` file by `typedef` instead simply define it in this place?

yosefe replied:
right, it should be in ucp,h, as as ucp_params_t

shssf commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,72 @@
+
```
Sorry, the `Copyright` missed.

shamisp replied:
well, at least mellanox copyright has to be here.

guy101 replied:
done

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2737/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2738/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/146/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/147/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2115/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2116/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2740/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/149/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2741/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/150/ for details.

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -639,6 +678,33 @@ void ucp_context_print_info(ucp_context_h context, FILE *stream);
+ * worker for each application thread, where every worker can be progressed
+ * independently of others.
+ *
+ * @note The worker object is allocated within context of the calling thread
```
after talking with @shssf  - need to mark move ucp_worker_create() as deprecated - see ucp_ep_destroy() as an example of how to do it.

shssf replied:
@yosefe May I propose a counter proposal? From the user point of view, It would be good to have ucp_worker_create() with default parameters in UCP interface. It will be the good example how to initialize parameters (need to move into ucp.h) and save the user time if the user need simple ucp_worker_create().
Proposal is:
- In `ucp_worker_create()` documentation, write that this is the same as `ucp_worker_create_param()` with default parameters.
- Move definition of the ucp_worker_create() from `.c` to `.h` to show the body of the function to the user (as example) and to the compiler (avoid extra jmp and expect inlining).
- If the user doesn't need parametrized function, ucp_worker_create() will be a good candidate.

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -639,6 +678,33 @@ void ucp_context_print_info(ucp_context_h context, FILE *stream);
+ *
+ * @param [in] context     Handle to @ref ucp_context_h
+ *                         "UCP application context".
+ * @param [in] worker_params User defined @ref worker_params_t "tunings" for the
```
it is not always just "tunings" - just call it "parameters for worker creation"

yosefe commented on src/ucp/api/ucp_def.h:
```diff
@@ -43,6 +43,14 @@ typedef struct ucp_tag_recv_info         ucp_tag_recv_info_t;
+ * @ingroup UCP_WORKER
+ * @brief UCP Worker parameters
+ *
+ * UCP worker params is a set of parameters passed to ucp_worker_create()
```
it's `@ref ucp_worker_create_with_params` and not ucp_worker_create()

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,81 @@
+
+/* Data structure to describe CPU mask.  */
+typedef struct
+{
```
bracket should on same line as "struct" (K&R)

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,81 @@
+        for ( _i = 0; _i <  (int)(UCS_CPU_SETSIZE / UCS_NCPUBITS); ++_i) {    \
+            ((_cpusetp)->ucs_bits)[_i] = 0;                                   \
+        };                                                                    \
+    } while (0);
```
no need to have ';' in the end

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,81 @@
+            |= UCS_CPUMASK (_cpu2))                                           \
+        :                                                                     \
+            0;                                                                \
+    } while(0);      
```
no need to have ';' in the end

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,81 @@
+            &= ~UCS_CPUMASK (_cpu2))                                          \
+        :                                                                     \
+            0;                                                                \
+    } while(0);
```
no need to have ';' in the end

guy101 replied:
it does not compile without ; after while(0)

yosefe replied:
; should be added when the macro is used and not when it's defined

yosefe commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,81 @@
+}
+
+#endif // UCS_TYPE_CPU_SET_H
+
```
too many space lines in the end

yosefe commented on src/uct/api/uct.h:
```diff
@@ -318,6 +319,7 @@ struct uct_iface_attr {
  * @ref uct_iface_open. User has to initialize all fields of this structure.
  */
 struct uct_iface_params {
+    ucs_cpu_set_t            cpu_set;     /** <cpu set to use while creating cq */
```
wrong doxygen-style comment, space before < and not after

yosefe commented on test/gtest/uct/uct_test.cc:
```diff
@@ -213,6 +213,7 @@ uct_test::entity::entity(const resource& resource, uct_iface_config_t *iface_con
     iface_params.tl_name     = const_cast<char*>(resource.tl_name.c_str());
     iface_params.dev_name    = const_cast<char*>(resource.dev_name.c_str());
     iface_params.rx_headroom = rx_headroom;
+    UCS_CPU_ZERO(&iface_params.cpu_set);
```
we need this in all other places calling uct_iface_open():
* test/examples/active_message.c
* src/tools/info/tl_info.c
* src/tools/perf/libperf.c

guy101 replied:
done

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2120/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2121/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2743/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2123/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/152/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2745/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/154/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2128/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2746/ for details.

mellanox-github commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/155/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2129/ for details (Mellanox internal link).

yosefe commented:
@shamisp pls review as well

yosefe commented:
bot:bgateNEW:retest

shamisp commented:
I have the same question that I raised during the call - what about memory affinity ?

shssf commented:
@shamisp We can do this in different PR. Too many things are already in this one.

shamisp commented on src/ucp/api/ucp.h:
```diff
@@ -639,6 +682,33 @@ void ucp_context_print_info(ucp_context_h context, FILE *stream);
+ *
+ * @return Error code as defined by @ref ucs_status_t
+ */
+ucs_status_t ucp_worker_create_p(ucp_context_h context,
```
what is "_P" stands for ?

shssf replied:
`p` - short from `parametrized`

shssf replied:
old name ucp_worker_create_with_parameters() looked too long. Propose the name please.

shamisp replied:
```_param ? ``` 
Or, can we bump up the version that preserve the original name ?

shamisp commented on src/ucp/core/ucp_worker.c:
```diff
@@ -171,7 +171,8 @@ static void ucp_worker_wakeup_context_cleanup(ucp_worker_wakeup_t *wakeup)
 static ucs_status_t ucp_worker_add_iface(ucp_worker_h worker,
-                                         ucp_rsc_index_t tl_id)
+                                         ucp_rsc_index_t tl_id,
+                                         ucs_cpu_set_t const * cpu_set_param)
```
I guess once we introduce memory affinity the interface will change again ? I suggest generalize the interface such that you path the arguments down the stack.

yosefe replied:
this is not interface - this is internal function

shamisp replied:
Still, why would not make it future proof ?

shamisp commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,77 @@
+        })                                                                              \
+
+
+static inline int uct_ib_find_lcs_cpu(ucs_cpu_set_t const *cpu_cores)
```
why IB ? what uct function is doing here ?

guy101 replied:
done

shamisp commented on src/ucs/type/cpu_set.h:
```diff
@@ -0,0 +1,77 @@
+#define UCS_CPU_SETSIZE  1024
+#define UCS_NCPUBITS     (8 * sizeof (ucs_cpu_mask_t))
+
+#define UCS_CPUELT(_cpu)  ((_cpu) / UCS_NCPUBITS)
```
http://stackoverflow.com/questions/25090635/use-and-in-c-programs

```
Here's what the C standard says (section 7.1.3):

All identifiers that begin with an underscore and either an uppercase letter or another underscore are always reserved for any use.
All identifiers that begin with an underscore are always reserved for use as identifiers with file scope in both the ordinary and tag name spaces.
```

yosefe replied:
The macro argument can be considered "with file scope"

yosefe replied:
also we defined in code style that macro arguments should begin with _ to avoid mix with "regular" identifiers

shamisp replied:
okay... in this case I think we are C compliant

guy101 replied:
done

shamisp commented on src/uct/api/uct.h:
```diff
@@ -318,6 +319,7 @@ struct uct_iface_attr {
  * @ref uct_iface_open. User has to initialize all fields of this structure.
  */
 struct uct_iface_params {
+    ucs_cpu_set_t            cpu_set;     /**< Cpu set to use while creating cq */
```
Cpu set  -> CPU set

guy101 replied:
why upper case? all other parameters are lower-case?

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/160/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2133/ for details (Mellanox internal link).

shamisp commented:
```_param``` ?
What happen with the original function ? Do we deprecate it ? If so, can we
just bump up the version number and keep the original name ?

On Wed, Nov 2, 2016 at 9:24 AM, shssf <notifications@github.com> wrote:

> *@shssf* commented on this pull request.
> ------------------------------
>
> In src/ucp/api/ucp.h <https://github.com/openucx/ucx/pull/1070>:
>
> > + * communication resources. For example, application can allocate a dedicated
> + * worker for each application thread, where every worker can be progressed
> + * independently of others.
> + *
> + * @note The worker object is allocated within context of the calling thread
> + *
> + * @param [in] context     Handle to @ref ucp_context_h
> + *                         "UCP application context".
> + * @param [in] params User defined @ref ucp_worker_params_t parameters
> + *                         for worker creation
> + * @param [out] worker_p   A pointer to the worker object allocated by the
> + *                         UCP library
> + *
> + * @return Error code as defined by @ref ucs_status_t
> + */
> +ucs_status_t ucp_worker_create_p(ucp_context_h context,
>
> old name ucp_worker_create_with_parameters() looked too long. Propose the
> name please.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openucx/ucx/pull/1070>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACIe2KmusEQpqwkI6XZNu0nHRErjio11ks5q6J0lgaJpZM4KcDGi>
> .
>

shssf commented:
@shamisp `param` is acceptable from my point of view. "Do we deprecate it ?" - originally, @yosefe proposed to deprecate this function. I made a counter proposal (nobody approved it) to do not deprecate it but expose to the user in `.h`. I described my proposal in one of the comment above.

yosefe commented:
let's rename to `ucp_worker_create_param`
i don't want to remove the old functions since it's in use, we can just comment that it's deprecated so new code will not use it.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2159/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2778/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/198/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2779/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/199/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2167/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2168/ for details (Mellanox internal link).

shamisp commented:
Before I approve it I want to hear about our versioning strategy.

shamisp commented:
This PR is waiting for #1125 to be merged

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2356/ for details (Mellanox internal link).

yosefe commented:
@guy101 @shamisp @bbenton pls take a look

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/439/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2932/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2371/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/440/ for details.

yosefe commented:
@shamisp pls take a look

yosefe commented:
@bbenton pls take a look - we are adding new flag in worker create API here

abouteiller commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -392,7 +393,8 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
 static ucs_status_t uct_ib_iface_create_cq(uct_ib_iface_t *iface, int cq_length,
-                                           size_t inl, struct ibv_cq **cq_p)
+                                           size_t inl, struct ibv_cq **cq_p,
+                                           int cq_comp_vector)
```
This variable name is cryptic a bit. Can it contain something else than a cpu_set mask?

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/457/ for details.

mellanox-github commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2947/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2406/ for details (Mellanox internal link).

yosefe commented:
@shamisp can we merge?

shamisp commented:
👍

shamisp commented:
you need @bbenton ok for this.

bbenton commented:
:+1: from an API/Spec perspective
